src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/FallbackProjectManager.cs

```diff
@@ -0,0 +1,163 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.IO;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem;
+
+/// <summary>
+/// This class is responsible for maintaining project information for projects that don't
+/// use the Razor or Web SDK, or otherwise don't get picked up by our CPS bits, but have
+/// .razor or .cshtml files regardless.
+/// </summary>
+[Shared]
+[Export(typeof(FallbackProjectManager))]
+internal sealed class FallbackProjectManager
+{
+    private readonly ProjectConfigurationFilePathStore _projectConfigurationFilePathStore;
+    private readonly LanguageServerFeatureOptions _languageServerFeatureOptions;
+    private readonly ProjectSnapshotManagerAccessor _projectSnapshotManagerAccessor;
+
+    private ImmutableHashSet<ProjectId> _fallbackProjectIds = ImmutableHashSet<ProjectId>.Empty;
+
+    [ImportingConstructor]
+    public FallbackProjectManager(
+        ProjectConfigurationFilePathStore projectConfigurationFilePathStore,
+        LanguageServerFeatureOptions languageServerFeatureOptions,
+        ProjectSnapshotManagerAccessor projectSnapshotManagerAccessor)
+    {
+        _projectConfigurationFilePathStore = projectConfigurationFilePathStore;
+        _languageServerFeatureOptions = languageServerFeatureOptions;
+        _projectSnapshotManagerAccessor = projectSnapshotManagerAccessor;
+    }
+
+    internal void DynamicFileAdded(ProjectId projectId, ProjectKey razorProjectKey, string projectFilePath, string filePath)
+    {
+        if (_fallbackProjectIds.Contains(projectId))
+        {
+            // If this is a fallback project, then Roslyn may not track documents in the project, so these dynamic file notifications
+            // are the only way to know about files in the project.
+            AddFallbackDocument(razorProjectKey, filePath, projectFilePath);
+        }
+        else if (_projectSnapshotManagerAccessor.Instance.GetLoadedProject(razorProjectKey) is null)
```

Comment:
Should `ProjectSnapshotManager` have a `TryGetLoadedProject` method?

Reply:
Not a bad idea. I'll make a note for follow up, to keep this change clean so it can go in 17.8

Reply:
Already working on that change. :smile: